### DISCUSSION TITLE
Records primary constructor - argument camel casing?

### DISCUSSION THREAD

**Discussion by davidroth** [dotnet/csharplang#3940] [General] [answered]
When defining a record using a primary constructor:

`public record Person(string Forename);`

This results in the following class (shortened):

```csharp
public class Person
{
   public Person(string Forename) => this.Forename = Forename;
   public Forename { get; init; }
}
```

Great. But it feels strange that we now have a PascalCased constructor parameter. (Forename)
Usually, C# method parameters are camelCased.

Question: Did the team consider automatically camelCasing constructor parameters when using primary constructors?

Ex.:

`public record Person(string Forename);`

would result in:

```csharp
public class Person
{
   public Person(string forename) => Forename = forename;
   public Forename { get; init; }
}
```

**Comment by Richiban** [comment, marked as answer]
I'd imagine that it's because changing case is not a trivial issue (it's not too bad if you assume English).

E.g. in English `big` and `BIG` are equal when you ignore case, but in Turkish they're not. Since unicode identifiers are allowed it's almost impossible to automatically change case on behalf of the programmer.

**Reply by HaloFour** [reply]
I recall this conversation happening fairly early on in the design of records but I can't find notes or comments about it.  Apart from the culture concerns with casing it's also not always obvious how acronyms or the like should be treated.  The .NET naming guidelines allow for capitalization of two-letter abbreviations, e.g. `IOFoo`.

**Reply by Joe4evr** [reply]
Somewhat relevant quote from Immo:
> Someone asked why our naming guidelines have to be so complex instead of simply recommending PascalCase everywhere.
> I answered "Because then `System.Io` would look like a namespace for manipulating a moon around Jupiter."

**Comment by YairHalberstadt** [comment]
You could achieve this using source generators. You write:

```csharp
partial record A(int myInt, string myString)
```

And the source generator generates:

```csharp
partial record A(int myInt, string myString){
    private int myInt { get; init; } = myInt;
    private string myString { get; init; } = myString;
    public int MyInt { get => myInt; init => myInt = value; }
    public string MyString { get => myString; init => myString = value; }
}
```

This works correctly for withers, equality, and printing prints `A { MyInt = 42, MyString = Hello World }`

**Comment by CyrusNajmabadi** [comment]
We discussed this at length in the ldm. Final conclusion was that it was entirely it if scope of the language to drive any casing policy or system here.

**Reply by HaloFour** [reply]
Was there a conversation with the CoreFX/BCL team?  I'm curious if they had any opinions as to how positional records might be adopted in their APIs, especially if they are the building blocks for discriminated unions.  Although I can imagine if the parameter/property naming is that much of a concern they could define them as nominal records with PascalCase property names and add constructors/destructors manually with camelCase parameter names.

**Reply by 333fred** [reply]
Several BCL members listen in regularly, and Stephen Toub is on the the language design team itself. I believe they were there for this discussion.
